control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -213,9 +213,11 @@ func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 				}
 
 				if hasBeenInjected(pod) {
-					endpointPods.Add(address.TargetRef.Name)
 					if isConsulDataplaneSupported(pod) {
-						if err = r.registerServicesAndHealthCheck(apiClient, pod, serviceEndpoints, healthStatus, endpointAddressMap); err != nil {
+						// Build the endpointAddressMap up for deregistering service instances later.
+						endpointAddressMap[pod.Status.PodIP] = true
```

Comment:
should this be after the registration call so that we know it was successful? (same below in `isGatway(pod)` case)

Reply:
Yeah, I like that more now that I have separated this from the registration functions.